python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py

```diff
@@ -0,0 +1,195 @@
+    assert isinstance(empty_index.index, _PDQHashIndex)
+    assert isinstance(empty_index.index.faiss_index, faiss.IndexFlatL2)
+    assert empty_index.index.faiss_index.d == DIMENSIONALITY
+    assert empty_index._deduper == dict()
+    assert empty_index._idx_to_entries == []
+
+
+def test_serialize_deserialize(empty_index) -> None:
+    buffer = io.BytesIO()
+    empty_index.serialize(buffer)
+    buffer.seek(0)
+    deserialized_index: PDQIndex2[str] = PDQIndex2.deserialize(buffer)
+
+    assert isinstance(deserialized_index, PDQIndex2)
+    assert deserialized_index.threshold == empty_index.threshold
+    assert isinstance(deserialized_index.index, _PDQHashIndex)
+    assert isinstance(deserialized_index.index.faiss_index, faiss.IndexFlatL2)
+    assert deserialized_index.index.faiss_index.d == DIMENSIONALITY
+    assert deserialized_index._deduper == empty_index._deduper
+    assert deserialized_index._idx_to_entries == empty_index._idx_to_entries
+
+
+def test_serialize_deserialize_with_custom_index_threshold(
+    custom_index_with_threshold,
+) -> None:
+    buffer = io.BytesIO()
+    custom_index_with_threshold.serialize(buffer)
+    buffer.seek(0)
+    deserialized_index: PDQIndex2[str] = PDQIndex2.deserialize(buffer)
+
+    assert isinstance(deserialized_index, PDQIndex2)
+    assert deserialized_index.threshold == custom_index_with_threshold.threshold
+    assert isinstance(deserialized_index.index, _PDQHashIndex)
+    assert isinstance(deserialized_index.index.faiss_index, faiss.IndexFlatL2)
+    assert deserialized_index.index.faiss_index.d == DIMENSIONALITY + 1
+    assert deserialized_index._deduper == custom_index_with_threshold._deduper
+    assert (
+        deserialized_index._idx_to_entries
+        == custom_index_with_threshold._idx_to_entries
+    )
+
+
+def test_empty_index_query(empty_index):
+    """Test querying an empty index."""
+    query_hash = "f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22"
+
+    # Query should return empty list
+    results = empty_index.query(query_hash)
+    assert len(results) == 0
+
+
+def test_sample_set_exact_match(sample_index):
+    """Test exact matches in sample set."""
+    index, pdq_hashes = sample_index
+
+    # Query with existing hash
+    query_hash = pdq_hashes[0]
+    results = index.query(query_hash)
+
+    assert len(results) == 1
+    assert (
+        results[0].similarity_info.distance == 0
+    )  # Exact match should have distance 0
+
+
+def test_sample_set_no_match(sample_index):
+    """Test no matches in sample set."""
+    index, _ = sample_index
+    results = index.query("b" * 64)
+    assert len(results) == 0
+
+
+def test_sample_set_near_match(sample_index):
+    """Test near matches in sample set."""
+    index, pdq_hashes = sample_index
+
+    # Create a near-match by flipping a few bits
+    base_hash = pdq_hashes[0]
+    near_hash = hex(int(base_hash, 16) ^ 0xF)[2:].zfill(64)  # Flip 4 bits
```

Comment:
blocking: Did you forget to query near_hash?

ignorable: In another similar project, I wrote a helper on the signal type itself to provide a random hash that was X distance away from a given hash by flipping random bits.

Reply:
I query the near_hash on the [line](https://github.com/facebook/ThreatExchange/pull/1698/files#diff-52467a9eead5544de4d6082c507dd5c96bebe25c34a1a5fdfebbdb9f2c223846R138) below